### PR TITLE
Added show_selection_panel.

### PR DESCRIPTION
**Pull request by Thom1729**
For #31.

I think the function is fine, or close to it. The tests, on the other hand, are a bit gnarly.

### CODE REVIEW

File: st3/sublime_lib/show_selection_panel.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+NOT_GIVEN = {}
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]:
```suggestion
NOT_GIVEN = object()
```

File: st3/sublime_lib/show_selection_panel.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    :argument on_highlight: A callback accepting a value from `items` to be
+    invoked every time the user changes the highlighted item in the panel.
+    """
+    labels_list = list(map(labels, items))
+
+    def on_done(index):
+        if index == -1:
+            if on_cancel:
+                on_cancel()
+        else:
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]:
```suggestion
        elif on_select:
```

And remove the next line and unindent the following. No idea how to submit multi-line suggestions.

File: st3/sublime_lib/show_selection_panel.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            if on_select:
+                on_select(items[index])
+
+    if selected == NOT_GIVEN:
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]:
```suggestion
    if selected is NOT_GIVEN:
```

File: st3/sublime_lib/show_selection_panel.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    :argument window: The :class:`sublime.Window` in which to show the panel.
+
+    :argument items: A :class:`list` of items to choose from.
+
+    Optional keyword arguments:
+
+    :argument flags: A bitwise OR of :const:`sublime.MONOSPACE_FONT` and
+    :const:`sublime.KEEP_OPEN_ON_FOCUS_LOST`.
+
+    :argument labels: A function taking elements of `items` to string labels.
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]: `items` (the parameter to `show_quick_panel`) may be a list of strings for multi-line display. Although we don't *have* to support it, it'd be a waste imo. So, a "label" can also be a list of strings, but their length should be consistent due to ST bugs.

Furthermore, I suggest an alternative use case where you already have the labels to show in the panel and the values to resolve to in a collection (or two). A list of two-tuples (or just two-item collections) sounds like a good compromise, where you'd have a list like follows:
```py
items = [
    ("label1", "value1"),
    ("label2", "value2"),
]
```
Allowing labels to be a list would be an easy way for `labels, values = zip(*items)` and then using these two lists in the `show_selection_panel` call.
Alternatively, passing a value->label map here might also seem useful, but that'd need some exploration. All values must be hashable in that case.

Also not so fond of the parameter name, if only functions are accepted. `labels` seems to imply it would expect a collection of labels, not a function that maps items to labels.

TL;DR
- Allow a list to be passed as the labels.
- Still allow a function to be passed. Useful for the default value of `str`, for example.
- Explore allowing a dict to be passed that has both values and labels; still need a separate param for order? value->label or label->value? different param?

File: st3/sublime_lib/show_selection_panel.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        raise ValueError("The items parameter must contain at least one item.")
+
+    if labels is NOT_GIVEN:
+        labels = items
+    elif callable(labels):
+        labels = list(map(labels, items))
+    elif len(items) != len(labels):
+        raise ValueError("The lengths of `items` and `labels` must match.")
+
+    if any(isinstance(label, list) for label in labels):
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]: What about allowing any sized collection that's not a string instead? (These are the situations where I wish strings weren't iterable.) Tuples satisfy that criteria, sets don't.

When tuples are allowed we can also used those in the various `tuple(map(...))` calls within this function, as tuples have less overhead due to being immutable.

File: st3/sublime_lib/show_selection_panel.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    :argument on_select: A callback accepting a value from `items` to be
+    invoked when the user chooses an item.
+
+    :argument on_cancel: A callback that will be invoked with no arguments if
+    the user closes the panel without choosing an item.
+
+    :argument on_highlight: A callback accepting a value from `items` to be
+    invoked every time the user changes the highlighted item in the panel.
+    """
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]: Add `:raises:` to docstrings? Doesn't need to be as detailed as to include every individual exception, but should at least mention all exception types. Also consider the exception that `list.index` throws.

File: st3/sublime_lib/show_selection_panel.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    elif callable(labels):
+        labels = list(map(labels, items))
+    elif len(items) != len(labels):
+        raise ValueError("The lengths of `items` and `labels` must match.")
+
+    if any(isinstance(label, list) for label in labels):
+        if not all(isinstance(label, list) for label in labels):
+            raise ValueError("Labels must be all strings or all lists.")
+
+        if len(set(map(len, labels))) != 1:
```
Comment (FichteFoll) [inline st3/sublime_lib/show_selection_panel.py]: I like this :+1:

**Comment by FichteFoll** [comment]
So yeah, no further changes requested but just a few things to think about.

Not sure when I'll manage to review your other PR. Maybe later today, maybe tomorrow.

File: st3/sublime_lib/show_selection_panel.py (hunk 5 above)
Comment (Thom1729) [reply st3/sublime_lib/show_selection_panel.py]: The underlying API takes lists, so we should pass those, but handling tuples ourselves is definitely a good idea.
